sklearn/metrics/pairwise.py

```diff
@@ -1271,6 +1271,13 @@ def pairwise_distances_chunked(X, Y=None, reduce_func=None,
             X_chunk = X[sl]
         D_chunk = pairwise_distances(X_chunk, Y, metric=metric,
                                      n_jobs=n_jobs, **kwds)
+        if ((X is Y or Y is None)
+                and PAIRWISE_DISTANCE_FUNCTIONS.get(metric, None)
+                is euclidean_distances):
+            # zeroing diagonal, taking care of aliases of "euclidean",
+            # i.e. "l2"
+            D_chunk[np.arange(D_chunk.shape[0]),
+                    np.arange(D_chunk.shape[1])[sl]] = 0
```

Comment:
surely there's a better way to force inner indexing?

Reply:
Can't we do `D_chunk.flat[offset::len(X)+1] = 0`?